devdoc/interlocked_requirements.md

```diff
@@ -0,0 +1,365 @@
+MOCKABLE_FUNCTION(, int32_t, interlocked_xor, volatile int32_t*, destination, int32_t, value);
+MOCKABLE_FUNCTION(, int16_t, interlocked_xor_16, volatile int16_t*, destination, int16_t, value);
+MOCKABLE_FUNCTION(, int64_t, interlocked_xor_64, volatile int64_t*, destination, int64_t, value);
+MOCKABLE_FUNCTION(, int8_t, interlocked_xor_8, volatile int8_t*, destination, int8_t, value);
+
+```
+
+## interlocked_add
+
+```c
+MOCKABLE_FUNCTION(, int32_t, interlocked_add, volatile int32_t*, addend, int32_t, value);
+```
+
+**SRS_INTERLOCKED_43_001 [** `interlocked_add` shall atomically add `*addend` with `value` and store the result in `addend`.**]**
+
+**SRS_INTERLOCKED_43_032: [** `interlocked_add` shall return the result of the addition.**]**
+
+## interlocked_and
+
+```c
+MOCKABLE_FUNCTION(, int32_t, interlocked_and, volatile int32_t*, destination, int32_t, value);
+```
+
+**SRS_INTERLOCKED_43_002 [** `interlocked_and` shall perform an atomic bitwise AND operation on the 32-bit integer values `*destination` and `value` and store the result in `*destination`.**]**
+
+**SRS_INTERLOCKED_43_033: [** `interlocked_and` shall return the initial value of `*destination`.**]**
+
+## interlocked_and_16
+
+```c
+MOCKABLE_FUNCTION(, int16_t, interlocked_and_16, volatile int16_t*, destination, int16_t, value);
+```
+
+**SRS_INTERLOCKED_43_003 [** `interlocked_and_16` shall perform an atomic bitwise AND operation on the 16-bit integer values `*destination` and `value` and store the result in `*destination.`**]**
+
+**SRS_INTERLOCKED_43_034: [** `interlocked_and_16` shall return the initial value of `*destination`.**]**
+
+## interlocked_and_64
+
+```c
+MOCKABLE_FUNCTION(, int64_t, interlocked_and_64, volatile int64_t*, destination, int64_t, value);
+```
+
+**SRS_INTERLOCKED_43_004 [** `interlocked_and_64` shall perform an atomic bitwise AND operation on the 64-bit integer values `*destination` and `value` and store the result in `*destination`.**]**
+
+**SRS_INTERLOCKED_43_035: [** `interlocked_and_64` shall return the initial value of `*destination`.**]**
+
+
+## interlocked_and_8
+
+```c
+MOCKABLE_FUNCTION(, int8_t, interlocked_and_8, volatile int8_t*, destination, int8_t, value);
+```
+
+**SRS_INTERLOCKED_43_005 [** `interlocked_and_8` shall perform an atomic bitwise AND operation on the 8-bit integer values `*destination` and `value` and store the result in `*destination`**]**
+
+**SRS_INTERLOCKED_43_036: [** `interlocked_and_8` shall return the initial value of `*destination`. **]**
+
+## interlocked_compare_exchange
+
+```c
+MOCKABLE_FUNCTION(, int32_t, interlocked_compare_exchange, volatile int32_t*, destination, int32_t, exchange, int32_t, comperand);
+```
+
+**SRS_INTERLOCKED_43_006 [** `interlocked_compare_exchange` shall compare the 32-bit integers pointed to by `destination` and `comperand`. If they are equal, `*destination` is set to `exchange`. These operations are performed atomically.***]**
```

Comment:
>If they are equal, `*destination` is set to `exchange` [](start = 139, length = 54)

This part for all the "compare and exchange" functions should be a separate spec (as Andrei mentioned, we should try to break down the specs into parts that can be tested).

Reply:
If I separate this spec into two, then the first spec is no longer testable,

1. `interlocked_compare_exchange` shall compare the 32-bit integers pointed to by `destination` and `comperand`.
2. If they are equal, `*destination` is set to `exchange`.

I don't think there is a way we can test if the comparison is happening.

---
In reply to: [425495832](https://github.com/Azure/azure-c-util/pull/39#discussion_r425495832) [](ancestors = 425495832)

Reply:
Personally I think having some "implicitly testable" requirements is ok. Like "shall compare x and y" is tested as part of the second requirement "if they are equal then...".

I kind of like just keeping the requirement statements shorter and more readable, without multiple clauses in a single requirement (you even have this with two sentences which feels like two specs to me).

But I'm okay with keeping it as one unit in this case if you want because I don't think we have any strict rules on what should be broken down or not in specs, it's just kind of guidelines and important to end up with good tests.

---
In reply to: [425866700](https://github.com/Azure/azure-c-util/pull/39#discussion_r425866700) [](ancestors = 425866700,425495832)